zypp/base/IOTools.cc

```diff
@@ -0,0 +1,129 @@
+/*---------------------------------------------------------------------\
+|                          ____ _   __ __ ___                          |
+|                         |__  / \ / / . \ . \                         |
+|                           / / \ V /|  _/  _/                         |
+|                          / /__ | | | | | |                           |
+|                         /_____||_| |_| |_|                           |
+|                                                                      |
+\---------------------------------------------------------------------*/
+/** \file	zypp/base/IOTools.cc
+ *
+*/
+
+#include <errno.h>
+#include <fcntl.h>
+#include <iostream>
+#include <glib.h>
+
+#include <zypp/AutoDispose.h>
+#include <zypp/base/IOTools.h>
+#include <zypp/base/Logger.h>
+
+namespace zypp::io {
+
+  BlockingMode setFILEBlocking (FILE * file, bool mode )
+  {
+    if ( !file ) return BlockingMode::FailedToSetMode;
+
+    int fd = ::fileno( file );
+
+    if ( fd == -1 )
+    { ERR << strerror( errno ) << std::endl; return BlockingMode::FailedToSetMode; }
+
+    int flags = ::fcntl( fd, F_GETFL );
+
+    if ( flags == -1 )
+    { ERR << strerror( errno ) << std::endl; return BlockingMode::FailedToSetMode; }
+
+    BlockingMode oldMode = ( flags & O_NONBLOCK ) == O_NONBLOCK ? BlockingMode::WasNonBlocking : BlockingMode::WasBlocking;
+    if ( !mode )
+      flags = flags | O_NONBLOCK;
+    else if ( flags & O_NONBLOCK )
+      flags = flags ^ O_NONBLOCK;
+
+    flags = ::fcntl( fd,F_SETFL,flags );
+
+    if ( flags == -1 )
+    { ERR << strerror(errno) << std::endl; return BlockingMode::FailedToSetMode; }
+
+    return oldMode;
+  }
+
+  std::pair<ReceiveUpToResult, std::string> receiveUpto(FILE *file, char c, timeout_type timeout, bool failOnUnblockError )
+  {
+    FILE * inputfile = file;
+    int    inputfileFd = ::fileno( inputfile );
+
+    size_t linebuffer_size = 0;
+    zypp::AutoFREE<char> linebuf;
+
+    const auto prevMode = setFILEBlocking( file, false );
+    if ( prevMode == BlockingMode::FailedToSetMode && failOnUnblockError )
+      return std::make_pair( ReceiveUpToResult::Error, std::string() );
+
+    // reset the blocking mode when we are done
+    zypp::OnScopeExit resetMode([ prevMode, fd = file ]( ){
+      if ( prevMode == BlockingMode::WasBlocking )
+        setFILEBlocking( fd, true );
+    });
+
+    bool haveTimeout = (timeout != no_timeout);
+    int remainingTimeout = static_cast<int>( timeout );
+    zypp::AutoDispose<GTimer *> timer;
```

Comment:
Better initialize to 0. AutoDispose does not 0-initialize integral types or ptr (unlike AutoFREE,FD,...).

Reply:
Done